jaxrs-service/src/main/java/com/quorum/tessera/p2p/PartyInfoResource.java

```diff
@@ -1,55 +1,143 @@
+        Client restClient,
+        Enclave enclave) {
+
+        this(partyInfoService, partyInfoParser, restClient, enclave, PayloadEncoder.create());
     }
 
     /**
      * Allows node information to be retrieved in a specific encoded form
      * including other node URLS and public key to URL mappings
      *
      * @param payload The encoded node information from the requester
-     * @return the merged node information from this node, which may contain new information
+     * @return the merged node information from this node, which may contain new
+     * information
      */
     @POST
     @Consumes(MediaType.APPLICATION_OCTET_STREAM)
     @Produces(MediaType.APPLICATION_OCTET_STREAM)
     @ApiOperation(value = "Request public key/url of other nodes", produces = "public keylist/url")
-    @ApiResponses({@ApiResponse(code = 200, message = "Encoded PartyInfo Data", response = byte[].class)})
+    @ApiResponses({
+        @ApiResponse(code = 200, message = "Encoded PartyInfo Data", response = byte[].class)})
     public Response partyInfo(@ApiParam(required = true) final byte[] payload) {
 
         final PartyInfo partyInfo = partyInfoParser.from(payload);
 
-        final PartyInfo updatedPartyInfo = partyInfoService.updatePartyInfo(partyInfo);
+        //Start validation stuff
+        PublicKey sender = enclave.defaultPublicKey();
+
+        final String url = partyInfo.getUrl();
+
+        final String dataToEncrypt = UUID.randomUUID().toString();
+
+        Predicate<Recipient> isValidRecipientKey = r -> {
+
+            PublicKey key = r.getKey();
+            final EncodedPayload encodedPayload = enclave.encryptPayload(dataToEncrypt.getBytes(), sender, Arrays.asList(key));
+
+            byte[] encodedPayloadData = payloadEncoder.encode(encodedPayload);
+
+            Response response = restClient.target(url)
+                .path("partyinfo")
+                .path("validate")
+                .request()
+                .post(Entity.entity(encodedPayloadData, MediaType.APPLICATION_OCTET_STREAM));
+            
+            String unencodedValidationData = response.readEntity(String.class);
+            
+            boolean isValid = Objects.equals(unencodedValidationData, dataToEncrypt);
+            if(!isValid) {
+                LOGGER.warn("Invalid key found {} recipient will be ignored.",url);
+            }
+            
+            return isValid;
+
+        };
+
+        Predicate<Recipient> isSendingUrl = r -> r.getUrl().equalsIgnoreCase(url);
+        
+        //Validate caller and treat no valid certs as security issue.  
+        partyInfo.getRecipients()
+            .stream()
+            .filter(isValidRecipientKey.and(isSendingUrl))
+            .findFirst()
+            .orElseThrow(() -> new SecurityException("No key found for url " + url));
+
+        final Set<Recipient> recipients = partyInfo.getRecipients()
+            .stream()
+            .filter(isValidRecipientKey)
```

Comment:
We should first filter on `isSendingUrl` to only try to check keys the sending node claimed to own, otherwise we will get a lot of expected misses.

Reply:
That’s the first check. The second is creating a filtered list of all recipients. If we don’t need to validate the recipients in party info then yes we can collect the initial filters

Reply:
The first check only checks that at least one valid key exists. The second is creating a filtered list of all recipients, to try and remove any keys that are not owned by the node that sent us the PartyInfo. Any key that is in the PartyInfo that is not owned by that node will, of course, not pass the check and get filtered - but the node never claimed to have owned that key in the first place, so why check with them?

Reply:
The first check only checks that at least one valid key exists. The second is creating a filtered list of all recipients, to try and remove any keys that are not owned by the node that sent us the PartyInfo. Any key that is in the PartyInfo that is not owned by that node will, of course, not pass the check and get filtered - but the node never claimed to have owned that key in the first place, so why check with them?